src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs

```diff
@@ -0,0 +1,373 @@
+            builder.Append(accountName);
+            builder.Append(address.AbsolutePath);
+            str.Append(builder);
+            var queryKeyValues = ExtractQueryKeyValues(address);
+            var dictionary = GetCommaSeparatedList(queryKeyValues);
+
+            foreach (var pair in dictionary.OrderBy(p => p.Key))
+            {
+                var stringBuilder = new StringBuilder(string.Empty);
+                stringBuilder.Append(pair.Key + ":");
+                var commaList = string.Join(",", pair.Value);
+                stringBuilder.Append(commaList);
+                str.Append("\n");
+                str.Append(stringBuilder);
+            }
+
+            return str.ToString();
+        }
+
+        /// <summary>
+        /// The get header values.
+        /// </summary>
+        /// <param name="headers">
+        /// The headers.
+        /// </param>
+        /// <param name="headerName">
+        /// The header name.
+        /// </param>
+        /// <returns>
+        /// The <see>
+        ///         <cref>List</cref>
+        ///     </see>
+        ///     .
+        /// </returns>
+        public static List<string> GetHeaderValues(WebHeaderCollection headers, string headerName)
+        {
+            var list = new List<string>();
+            var values = headers[headerName];
+            if (values != null)
+            {
+                list.Add(values.TrimStart(null));
+            }
+
+            return list;
+        }
+
+        /// <summary>
+        /// The extract query key values.
+        /// </summary>
+        /// <param name="address">
+        /// The address.
+        /// </param>
+        /// <returns>
+        /// The <see>
+        ///         <cref>Dictionary</cref>
+        ///     </see>
+        ///     .
+        /// </returns>
+        private static Dictionary<string, HashSet<string>> ExtractQueryKeyValues(Uri address)
+        {
+            var values = new Dictionary<string, HashSet<string>>();
+            var newreg = new Regex(@"\?(\w+)\=([\w|\=]+)|\&(\w+)\=([\w|\=]+)");
+            var matches = newreg.Matches(address.Query);
+            foreach (Match match in matches)
+            {
+                string key, value;
+                if (!string.IsNullOrEmpty(match.Groups[1].Value))
+                {
+                    key = match.Groups[1].Value;
+                    value = match.Groups[2].Value;
+                }
+                else
+                {
+                    key = match.Groups[3].Value;
+                    value = match.Groups[4].Value;
+                }
+
+                if (!values.ContainsKey(key))
+                {
+                    var newSet = new HashSet<string> { value };
+                    values.Add(key, newSet);
+                }
+                else
+                {
+                    HashSet<string> setOfValues = null;
+                    values.TryGetValue(key, out setOfValues);
```

Comment:
This "try" should be promoted up to the "if". At the moment, you test, and then test a second time.